rolling upgrade of application fails in embedded mode

<!--
Thanks for reporting your issue. Please share with us the following information, to help us resolve your issue quickly and efficiently.
-->
  
**Describe the bug**
Application has embedded hazelcast.
Application is upgrading in rolling fashion from V1 to V2.
V2 introduced a new cache (V2-cache) with new value type (V2-classDef1)


1.Let's say I have four nodes in cluster: 4 - V1 nodes
2. Rolling upgrade is triggered.
3. Currently it 3 V1 nodes and 1 V2 node forming hazelcast cluster. 
At this point V2 created V2-Cache.
4.For some application specific reason upgrade has to rollback.
5.So V2 is shutdown. V2-Cache cannot be deleted.
6. New V1 is triggered 
7. New V1 fails saying it could not find "V2-classDef1" as cache is present in cluster.


**Expected behavior**

How to allow V1 HZ instance to startup without failing though any where in cluster class definition is not present?
Justification: Failure makes sense when some one accesses that new cache.

**To Reproduce**

Steps to reproduce the behavior:
1. Go to '...'
2. Click on '....'
3. ...

**Additional context**

<!--
Add any other context about the problem here.

Common details that we're often interested in:
- Detailed description of the steps to reproduce your issue
- Logs and stack traces, if available
- Hazelcast version that you use (e.g. 3.4, also specify whether it is a minor release or the latest snapshot)
- If available, integration module versions (e.g. Tomcat, Jetty, Spring, Hibernate). Also, include their detailed configuration information such as web.xml, Hibernate configuration and `context.xml` for Spring
- Cluster size, i.e. the number of Hazelcast cluster members
- Number of the clients
- Version of Java. It is also helpful to mention the JVM parameters
- Operating system. If it is Linux, kernel version is helpful
- Unit test with the `hazelcast.xml` file. If you could include a unit test which reproduces your issue, we would be grateful
-->

Thanks  for your quick response @vbekiaris 
We are doing pretty straightforward way.


**Code that creates cache in V2 node**

private Cache createCache(String name, Class valueType) {
CacheManager cacheManager = HazelcastServerCachingProvider.createCachingProvider(gethz()).getCacheManager();
Cache cache = cacheManager.createCache(name, createCacheConfig(name, valueType));
return cache;
}
private CacheConfig createCacheConfig(String name,Class valueType ) {
CacheConfig cfg = new CacheConfig();
cfg.setTypes(String.class, valueType);
return cfg.setName(name);
}

**Exception Seen on V1 node at Step 7 mentioned above**

om.hazelcast.internal.cluster.impl.operations.JoinRequestOp
SEVERE: [55.55.55.5]:5701 [dev] [3.11.5] Could not resolve type com.pega.ClassA
com.hazelcast.core.HazelcastException: Could not resolve type com.pega.ClassA
at com.hazelcast.config.AbstractCacheConfig.resolve(AbstractCacheConfig.java:350)
at com.hazelcast.config.AbstractCacheConfig.resolveValueType(AbstractCacheConfig.java:340)
at com.hazelcast.config.AbstractCacheConfig.getValueType(AbstractCacheConfig.java:322)
at com.hazelcast.config.CacheConfig.copy(CacheConfig.java:682)
at com.hazelcast.cache.impl.PreJoinCacheConfig.<init>(PreJoinCacheConfig.java:59)
at com.hazelcast.cache.impl.PreJoinCacheConfig.<init>(PreJoinCacheConfig.java:55)
at com.hazelcast.cache.impl.AbstractCacheService.getPreJoinOperation(AbstractCacheService.java:656)
at com.hazelcast.spi.impl.NodeEngineImpl.getPreJoinOperations(NodeEngineImpl.java:451)
at com.hazelcast.internal.cluster.impl.ClusterJoinManager.preparePreJoinOps(ClusterJoinManager.java:777)
at com.hazelcast.internal.cluster.impl.ClusterJoinManager.checkIfJoinRequestFromAnExistingMember(ClusterJoinManager.java:649)
at com.hazelcast.internal.cluster.impl.ClusterJoinManager.checkJoinRequest(ClusterJoinManager.java:267)
at com.hazelcast.internal.cluster.impl.ClusterJoinManager.executeJoinRequest(ClusterJoinManager.java:247)
at com.hazelcast.internal.cluster.impl.ClusterJoinManager.handleJoinRequest(ClusterJoinManager.java:171)
at com.hazelcast.internal.cluster.impl.operations.JoinRequestOp.run(JoinRequestOp.java:41)
at com.hazelcast.spi.Operation.call(Operation.java:170)
at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:208)
at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:197)
at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:413)
at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:153)
at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:123)
at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:110)
Caused by: java.lang.ClassNotFoundException: com.pega.ClassA
at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
at com.hazelcast.nio.ClassLoaderUtil.tryLoadClass(ClassLoaderUtil.java:288)
at com.hazelcast.nio.ClassLoaderUtil.loadClass(ClassLoaderUtil.java:252)
at com.hazelcast.config.AbstractCacheConfig.resolve(AbstractCacheConfig.java:348)
... 20 more

--- a/hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java
+++ b/hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java
@@ -762,7 +762,7 @@ public abstract class AbstractCacheService implements ICacheService, PreJoinAwar
         OnJoinCacheOperation preJoinCacheOperation;
         preJoinCacheOperation = new OnJoinCacheOperation();
         for (Map.Entry<String, CompletableFuture<CacheConfig>> cacheConfigEntry : configs.entrySet()) {
-            CacheConfig cacheConfig = new PreJoinCacheConfig(cacheConfigEntry.getValue().join());
+            CacheConfig cacheConfig = new PreJoinCacheConfig(cacheConfigEntry.getValue().join(), false);
             preJoinCacheOperation.addCacheConfig(cacheConfig);
         }
         return preJoinCacheOperation;

--- a/hazelcast/src/test/java/com/hazelcast/cache/CacheTypesConfigTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cache/CacheTypesConfigTest.java
@@ -26,6 +26,7 @@ import com.hazelcast.config.UserCodeDeploymentConfig;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.internal.util.FilteringClassLoader;
 import com.hazelcast.internal.util.RootCauseMatcher;
+import com.hazelcast.spi.properties.ClusterProperty;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.TestHazelcastInstanceFactory;
@@ -88,6 +89,24 @@ public class CacheTypesConfigTest extends HazelcastTestSupport {
         expect.expectCause(new RootCauseMatcher(ClassNotFoundException.class, "classloading.domain.PersonEntryProcessor - "
                 + "Package excluded explicitly"));
         cache.invoke(key, new PersonEntryProcessor());
+    }
+
+    @Test
+    public void cacheConfigShouldBeAddedOnJoiningMember_whenNoMemberResolvesClass() {
+        // given a member who cannot resolve the value type of a CacheConfig
+        HazelcastInstance hz1 = factory.newHazelcastInstance(getClassFilteringConfig());
+        // and a new member that creates a CacheConfig with an explicit value type
+        HazelcastInstance hz2 = factory.newHazelcastInstance(getConfig());
+        CachingProvider cachingProvider = createServerCachingProvider(hz1);
+        cachingProvider.getCacheManager().createCache(cacheName, createCacheConfig());
+        // ensure cluster is formed but cache is not used
+        assertClusterSize(2, hz1, hz2);
+        // member that is aware of value type leaves cluster
+        hz2.shutdown();
+
+        // then new member unaware of the value type can join the cluster
+        hz2 = factory.newHazelcastInstance(getClassFilteringConfig());
+        assertClusterSize(2, hz1, hz2);
     }
 
     // When the joining member is not aware of key or value class but it is later resolvable via user code deployment, then
@@ -162,6 +181,12 @@ public class CacheTypesConfigTest extends HazelcastTestSupport {
         CacheConfig<String, Person> cacheConfig = new CacheConfig<>();
         cacheConfig.setTypes(String.class, Person.class);
         return cacheConfig;
+    }
+
+    @Override
+    protected Config getConfig() {
+        return smallInstanceConfig()
+                .setProperty(ClusterProperty.MAX_JOIN_SECONDS.getName(), "10");
     }
 
     private Config getClassFilteringConfig() {

